### ISSUE TITLE
 AzureSpringBootRequestHandler does not manage context lock in case of concurent calls

### ISSUE THREAD

**Issue by alain-marcel** [spring-cloud/spring-cloud-function#526]
Hello,
AzureSpringBootRequestHandler class  (of  spring-cloud-function-adapter-azure) does not manage concurenct calls.
Symptom: when calling multiple functions at same time of a same AzureFunction (all written using AzureSpringBootRequestHandler),
one function may be called instead of another.

Find bellow an implementation example to solve this issue
  - AzureSpringFunctionAdapter extends AbstractSpringFunctionAdapterInitializer
  - AzureFunctions : contains two ConcurrentHashMap to manage concurency
  - Miscellaneous classes

it wolud be great if issue will be remove from next version.
Regards
Alain

======================================================= AzureSpringFunctionAdapter.java
package example.spring.cloud.function.azure.adapter;

import org.springframework.cloud.function.utils.FunctionClassUtils;

public final class AzureSpringFunctionAdapter$ {
   public static final AzureSpringFunctionAdapter$ MODULE$ = new AzureSpringFunctionAdapter$();

   public AzureSpringFunctionAdapter apply(final String functionName) {
      return new AzureSpringFunctionAdapter(functionName, FunctionClassUtils.getStartClass());
   }

   private AzureSpringFunctionAdapter$() {
   }
}

package example.spring.cloud.function.azure.adapter;

import com.microsoft.azure.functions.ExecutionContext;
import com.microsoft.azure.functions.OutputBinding;
import java.util.Collection;
import java.util.Objects;
import java.util.function.Function;
import org.reactivestreams.Publisher;
import org.springframework.cloud.function.context.AbstractSpringFunctionAdapterInitializer;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

public class AzureSpringFunctionAdapter extends AbstractSpringFunctionAdapterInitializer {
   private final String functionName;

   public void initialize(final ExecutionContext targetContext) {
      super.initialize(targetContext);
   }

   public void close() {
      super.close();
   }

   public Object handleRequest(final ExecutionContext context) {
      return this.handleRequest((Object)null, context);
   }

   public Object handleRequest(final Object input, final ExecutionContext context) {
      Objects.requireNonNull(context);

      try {
         Publisher events = input == null ? Mono.empty() : this.extract(this.convertEvent(input));
         Publisher output = this.apply((Publisher)events);
         Object res = this.result(input, output);
         return res;
      } catch (Throwable var7) {
         throw new Exception(var7);
      }
   }

   public String doResolveName(final Object targetContext) {
      return this.functionName;
   }

   public void handleOutput(final Object input, final OutputBinding binding, final ExecutionContext context) {
      Object result = this.handleRequest(input, context);
      binding.setValue(result);
   }

   public Object convertEvent(final Object input) {
      return input;
   }

   public Flux extract(final Object input) {
      return !this.isSingleInput(this.getFunction(), input) ? Flux.fromIterable((Iterable)input) : Flux.just(input);
   }

   public boolean isSingleInput(final Function function, final Object input) {
      if (!(input instanceof Collection)) {
         return true;
      } else if (this.getInspector() != null) {
         return Collection.class.isAssignableFrom(this.getInspector().getInputType(function));
      } else {
         return ((Collection)input).size() <= 1;
      }
   }

   public boolean isSingleOutput(final Function function, final Object output) {
      if (!(output instanceof Collection)) {
         return true;
      } else if (this.getInspector() != null) {
         return Collection.class.isAssignableFrom(this.getInspector().getOutputType(function));
      } else {
         return ((Collection)output).size() <= 1;
      }
   }

   public void initialize(final Object targetContext) {
      this.initialize((ExecutionContext)targetContext);
   }

   public AzureSpringFunctionAdapter(final String functionName, final Class configurationClass) {
      super(configurationClass);
      this.functionName = functionName;
   }
}

======================================================= AzureFunctions.java
package example.spring.cloud.function.azure.adapter;

public final class AzureFunctions$ {
   public static final AzureFunctions$ MODULE$ = new AzureFunctions$();
   private static final AzureFunctions instance = new AzureFunctions();

   public AzureFunctions instance() {
      return instance;
   }

   private AzureFunctions$() {
   }
}

package example.spring.cloud.function.azure.adapter;

import com.microsoft.azure.functions.ExecutionContext;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

public class AzureFunctions {
   private final ConcurrentHashMap adaptersByFunctionName;
   private final ConcurrentHashMap initializedByFunctionName;

   public static AzureFunctions instance() {
      return AzureFunctions$.MODULE$.instance();
   }

   private ConcurrentHashMap adaptersByFunctionName() {
      return this.adaptersByFunctionName;
   }

   private ConcurrentHashMap initializedByFunctionName() {
      return this.initializedByFunctionName;
   }

   public AzureSpringFunctionAdapter getOrInit(final ExecutionContext executionContext) {
      Objects.requireNonNull(executionContext);
      String functionName = (String)Objects.requireNonNull(executionContext.getFunctionName());
      this.initializedByFunctionName().computeIfAbsent(functionName, (x$1) -> {
         return AzureFunctionInitializationStatus.INITIALIZATION_IN_PROGRESS;
      });
      AzureSpringFunctionAdapter adapter = (AzureSpringFunctionAdapter)this.adaptersByFunctionName().computeIfAbsent(functionName, (funcName) -> {
         return this.initializeAdapter(funcName, executionContext);
      });
      this.initializedByFunctionName().put(functionName, AzureFunctionInitializationStatus.INITIALIZED);
      return adapter;
   }

   public void init(final String functionName) {
      if (!this.initializedByFunctionName().containsKey(functionName)) {
         this.getOrInit(new BasicInitializationContext(functionName));
      }

   }

   private AzureSpringFunctionAdapter initializeAdapter(final String funcName, final ExecutionContext executionContext) {
      AzureSpringFunctionAdapter adapter = com.lpg.sc.util_213.spring.cloud.function.azure.adapter.AzureSpringFunctionAdapter..MODULE$.apply(funcName);
      adapter.initialize(executionContext);
      return adapter;
   }

   public AzureFunctions() {
      this.adaptersByFunctionName = new ConcurrentHashMap();
      this.initializedByFunctionName = new ConcurrentHashMap();
   }
}

             

======================================================= AzureFunctionInitializationStatus.java
package example.spring.cloud.function.azure.adapter;

public enum AzureFunctionInitializationStatus {
    NOT_INITIALIZED,
    INITIALIZATION_IN_PROGRESS,
    INITIALIZED
}

======================================================= BasicInitializationContext.java
package example.spring.cloud.function.azure.adapter;

import com.microsoft.azure.functions.ExecutionContext;
import java.util.UUID;
import java.util.logging.Logger;

public class BasicInitializationContext implements ExecutionContext {
   private final String functionName;
   private final String invocationId;

   public String functionName() {
      return this.functionName;
   }

   public String invocationId() {
      return this.invocationId;
   }

   public Logger getLogger() {
      return Logger.getAnonymousLogger();
   }

   public String getInvocationId() {
      return this.invocationId();
   }

   public String getFunctionName() {
      return this.functionName();
   }

   public BasicInitializationContext(final String functionName) {
      this.functionName = functionName;
      this.invocationId = UUID.randomUUID().toString();
   }
}

**Comment by pfemeiros** [comment]
I have the same issue. I identified that it started to occur after version 2.0.2.RELEASE. Apparentely it occur because every time you call a different function, a new spring cloud application starts.

To exemplify: https://github.com/pfemeiros/function-sample-azure
The branch named **spring_cloud_2.0.2** have no issue running in parallel.
The branch named **spring_cloud_3.0.8** have the described problem.

Regards
Pedro

**Comment by olegz** [comment]
I am going to close this as it was just addressed with https://github.com/spring-cloud/spring-cloud-function/commit/b0ef4522d48a7b29aeeb3ef9c0d3ca939bc87b5f, so I believe it's a duplicate.

Feel free to raise a new issue or re-open this one if problem persists.
